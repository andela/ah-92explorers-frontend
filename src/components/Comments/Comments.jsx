/* eslint-disable react/no-unused-state */
/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
import React from 'react';
import moment from 'moment';
import { confirmAlert } from 'react-confirm-alert';
import '../../assets/auth/css/style.css';
import { connect } from 'react-redux';
import { sortArrayDesd as sortComments } from 'tesla-error-handler';
import manIcon from '../../assets/images/man.jpg';
import { fetchComments, postComment, deleteComment } from '../../redux/actions/actionCreators/comments';
import '../../assets/css/comment.css';
import 'react-confirm-alert/src/react-confirm-alert.css';

export class Comments extends React.Component {
  state = {
    body: '',
  }

  componentDidMount() {
    const { slug } = this.props;
    this.props.fetchComments(slug);
  }

  handleOnComment = (body, slug) => {
    this.props.postComment(body, slug).then((res) => {
      this.props.fetchComments(slug);
      this.setState({ body: '' });
    });
  }

  onChange = (e) => {
    this.setState({ body: e.target.value });
  }

  onDeleteComment = itemId => {
    confirmAlert({
      title: 'Confirm to delete',
      message: 'Are you sure, you want to delete this comment?',
      buttons: [
        {
          label: 'Yes',
          onClick: () => {
            this.props.deleteComment(itemId).then((res) => {
              const { slug } = this.props;
              this.props.fetchComments(slug);
            });
          },
        },
        {
          label: 'No',
          onClick: () => {
          },
        },
      ],
    });
  }

  render() {
    const {
      username, userImage, slug,
    } = this.props;
    const { body } = this.state;
    const { comments } = this.props;
    return (
      <div>
        <div className="commentForm">
          <div className="profileSection">
            <img src={userImage} alt={username} className="commPic" />
            <span className="profileName">{username}</span>
          </div>
          <div className="comment-field">
            <input
              type="text"
              value={body}
              name="body"
              className="cmBox"
              placeholder="Enter a comment here"
              onChange={this.onChange}
            />
          </div>
          <br />
          <button className="publish" type="button" onClick={() => this.handleOnComment(body, slug)}>Comment</button>
        </div>

        <div className="comment-list">
          {sortComments(comments).map(element => (
            <div className="commentCard" key={element.id}>
              <div className="commentSection">
                <img src={element.commentor.image ? element.commentor.image : manIcon} alt={element.commentor.username} className="commPic" />
                <span className="profileName">{element.commentor.username}</span>
                <div className="comment-date">
                  {' '}
                  {moment(element.updatedAt).fromNow()}
                </div>
                <p className="comment-body">
                  {element.body}
                </p>
              </div>
              <div className="commentFooter">
                { element.commentor.username === localStorage.getItem('username')
                  ? (
                    <img
                      src={require('../../assets/icons/trash.svg')}
                      alt=""
                      className="bodyIcons footerIconsRight trash"
                      onClick={() => this.onDeleteComment(element.id)}
                    />
                  )
                  : ''}
              </div>
            </div>
          ))}
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  comments: state.comments.comments,
});

export default connect(
  mapStateToProps, { fetchComments, postComment, deleteComment },
)(Comments);
