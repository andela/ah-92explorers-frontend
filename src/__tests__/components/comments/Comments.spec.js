import '@babel/polyfill';
import React from 'react';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Comments } from '../../../components/Comments/Comments.jsx';

describe('<Comments />', () => {
  const props = {
    comments: [],
    slug: '',
    username: '',
    userImage: '',
    body: '',
    fetchComments: () => jest.fn(),
    postComment: async () => {},
    deleteComment: () => jest.fn(),
  };

  const initialState = {};
  const mockStore = configureStore();
  let store;

  store = mockStore(initialState);
  const wrapper = mount(<Comments store={store} {...props}/>);

  it('should render Comments component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render one main div element', () => {
    expect(wrapper.find('div')).toHaveLength(5);
  });

  it('shoudl handle on comment', async () => {
    const fakeEvent = { preventDefault: () => {} };
    const component = wrapper.instance();
    component.handleOnComment(fakeEvent);
    expect(component).toBeDefined();
  });

  it('should handle on change', () => {
    const component = wrapper.instance();
    const fakeEvent = { target: { name: 'title', value: 'THIS IS TITLE' } };
    component.onChange(fakeEvent);
    expect(component).toBeDefined();
  });

  it('should handle on click on delete comment', () => {
    const fakeEvent = { preventDefault: () => {} };
    const component = wrapper.instance();
    component.onDeleteComment(fakeEvent);
    expect(component).toBeDefined();
  });

  it('hould simulate on click comment button', async () => {
    const wrapper = shallow(<Comments { ...props} />);
    const btn = wrapper.find('.publish');
    btn.simulate('click', { preventDefault: jest.fn() });
    expect(btn.length).toBe(1);
  });

  it('should find the key', async () => {
    const wrapper = mount(<Comments { ...props} />);
    const btn = wrapper.find('.commentCard');
    expect(btn).toBeDefined();
  });
});