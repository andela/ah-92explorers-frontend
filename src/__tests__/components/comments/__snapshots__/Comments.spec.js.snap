// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Comments /> should render Comments component 1`] = `ReactWrapper {}`;
